In [1]:
import pandas as pd
from scipy.io import loadmat
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import NearMiss 

In [2]:
liverPatientData = loadmat('../MATLAB Code/LiverPatientData.mat')

liverPatientData

{'__header__': b'MATLAB 5.0 MAT-file, Platform: MACI64, Created on: Sat Jan  9 17:32:18 2021',
 '__version__': '1.0',
 '__globals__': [],
 'liverPatientData': array([[65.  ,  2.  ,  0.7 , ...,  3.3 ,  0.9 ,  0.  ],
        [62.  ,  1.  , 10.9 , ...,  3.2 ,  0.74,  0.  ],
        [62.  ,  1.  ,  7.3 , ...,  3.3 ,  0.89,  0.  ],
        ...,
        [52.  ,  1.  ,  0.8 , ...,  3.2 ,  1.  ,  0.  ],
        [31.  ,  1.  ,  1.3 , ...,  3.4 ,  1.  ,  0.  ],
        [38.  ,  1.  ,  1.  , ...,  4.4 ,  1.5 ,  1.  ]])}

In [3]:
# The code below needs to be commented out after executing once.
liverPatientData = liverPatientData['liverPatientData']

liverPatientData

array([[65.  ,  2.  ,  0.7 , ...,  3.3 ,  0.9 ,  0.  ],
       [62.  ,  1.  , 10.9 , ...,  3.2 ,  0.74,  0.  ],
       [62.  ,  1.  ,  7.3 , ...,  3.3 ,  0.89,  0.  ],
       ...,
       [52.  ,  1.  ,  0.8 , ...,  3.2 ,  1.  ,  0.  ],
       [31.  ,  1.  ,  1.3 , ...,  3.4 ,  1.  ,  0.  ],
       [38.  ,  1.  ,  1.  , ...,  4.4 ,  1.5 ,  1.  ]])

In [4]:
liverPatientData = pd.DataFrame(liverPatientData)

liverPatientData.columns = ['Age', 'Gender', 'Total Bilirubin', 
                            'Direct Bilirubin', 'Alkaline Phosphotase', 'Alamine Aminotransferase', 
                            'Aspartate Aminotransferase', 'Total Protiens', 'Albumin', 
                            'Ratio Albumin and Globulin Ratio', 'Liver Patient']

liverPatientData.head(15)

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0   65.0     2.0              0.7               0.1                 187.0   
1   62.0     1.0             10.9               5.5                 699.0   
2   62.0     1.0              7.3               4.1                 490.0   
3   58.0     1.0              1.0               0.4                 182.0   
4   72.0     1.0              3.9               2.0                 195.0   
5   46.0     1.0              1.8               0.7                 208.0   
6   26.0     2.0              0.9               0.2                 154.0   
7   29.0     2.0              0.9               0.3                 202.0   
8   17.0     1.0              0.9               0.3                 202.0   
9   55.0     1.0              0.7               0.2                 290.0   
10  57.0     1.0              0.6               0.1                 210.0   
11  72.0     1.0              2.7               1.3                 260.0   
12  64.0     1.0              0.9               0.3                 310.0   
13  74.0     2.0              1.1               0.4                 214.0   
14  61.0     1.0              0.7               0.2                 145.0   

    Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                       16.0                        18.0             6.8   
1                       64.0                       100.0             7.5   
2                       60.0                        68.0             7.0   
3                       14.0                        20.0             6.8   
4                       27.0                        59.0             7.3   
5                       19.0                        14.0             7.6   
6                       16.0                        12.0             7.0   
7                       14.0                        11.0             6.7   
8                       22.0                        19.0             7.4   
9                       53.0                        58.0             6.8   
10                      51.0                        59.0             5.9   
11                      31.0                        56.0             7.4   
12                      61.0                        58.0             7.0   
13                      22.0                        30.0             8.1   
14                      53.0                        41.0             5.8   

    Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0       3.3                              0.90            0.0  
1       3.2                              0.74            0.0  
2       3.3                              0.89            0.0  
3       3.4                              1.00            0.0  
4       2.4                              0.40            0.0  
5       4.4                              1.30            0.0  
6       3.5                              1.00            0.0  
7       3.6                              1.10            0.0  
8       4.1                              1.20            1.0  
9       3.4                              1.00            0.0  
10      2.7                              0.80            0.0  
11      3.0                              0.60            0.0  
12      3.4                              0.90            1.0  
13      4.1                              1.00            0.0  
14      2.7                              0.87            0.0

In [5]:
liverPatientData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 579 entries, 0 to 578
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age                               579 non-null    float64
 1   Gender                            579 non-null    float64
 2   Total Bilirubin                   579 non-null    float64
 3   Direct Bilirubin                  579 non-null    float64
 4   Alkaline Phosphotase              579 non-null    float64
 5   Alamine Aminotransferase          579 non-null    float64
 6   Aspartate Aminotransferase        579 non-null    float64
 7   Total Protiens                    579 non-null    float64
 8   Albumin                           579 non-null    float64
 9   Ratio Albumin and Globulin Ratio  579 non-null    float64
 10  Liver Patient                     579 non-null    float64
dtypes: float64(11)
memory usage: 49.9 KB


In [6]:
inputs = liverPatientData.iloc[:, 0:10]
target = liverPatientData.iloc[:, 10]

In [7]:
inputs.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  65.0     2.0              0.7               0.1                 187.0   
1  62.0     1.0             10.9               5.5                 699.0   
2  62.0     1.0              7.3               4.1                 490.0   
3  58.0     1.0              1.0               0.4                 182.0   
4  72.0     1.0              3.9               2.0                 195.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      16.0                        18.0             6.8   
1                      64.0                       100.0             7.5   
2                      60.0                        68.0             7.0   
3                      14.0                        20.0             6.8   
4                      27.0                        59.0             7.3   

   Albumin  Ratio Albumin and Globulin Ratio  
0      3.3                              0.90  
1      3.2                              0.74  
2      3.3                              0.89  
3      3.4                              1.00  
4      2.4                              0.40

In [8]:
target.head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Liver Patient, dtype: float64

In [9]:
# Original split.
# We set the training size to 405 as this is what the split was in MATLAB for the original MLP.
X_train, X_test_first, y_train, y_test_first = train_test_split(inputs, target, train_size=405, 
                                                    random_state=1)


X_val, X_test, y_val, y_test = train_test_split(X_test_first, y_test_first, test_size=0.5, 
                                                    random_state=1)

In [10]:
X_train.info() # We should have 405 observations.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 405 entries, 370 to 37
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age                               405 non-null    float64
 1   Gender                            405 non-null    float64
 2   Total Bilirubin                   405 non-null    float64
 3   Direct Bilirubin                  405 non-null    float64
 4   Alkaline Phosphotase              405 non-null    float64
 5   Alamine Aminotransferase          405 non-null    float64
 6   Aspartate Aminotransferase        405 non-null    float64
 7   Total Protiens                    405 non-null    float64
 8   Albumin                           405 non-null    float64
 9   Ratio Albumin and Globulin Ratio  405 non-null    float64
dtypes: float64(10)
memory usage: 34.8 KB


In [11]:
X_val.info() # We should have 87 observations.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87 entries, 346 to 41
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age                               87 non-null     float64
 1   Gender                            87 non-null     float64
 2   Total Bilirubin                   87 non-null     float64
 3   Direct Bilirubin                  87 non-null     float64
 4   Alkaline Phosphotase              87 non-null     float64
 5   Alamine Aminotransferase          87 non-null     float64
 6   Aspartate Aminotransferase        87 non-null     float64
 7   Total Protiens                    87 non-null     float64
 8   Albumin                           87 non-null     float64
 9   Ratio Albumin and Globulin Ratio  87 non-null     float64
dtypes: float64(10)
memory usage: 7.5 KB


In [12]:
X_test.info() # We should have 87 observations.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87 entries, 574 to 560
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age                               87 non-null     float64
 1   Gender                            87 non-null     float64
 2   Total Bilirubin                   87 non-null     float64
 3   Direct Bilirubin                  87 non-null     float64
 4   Alkaline Phosphotase              87 non-null     float64
 5   Alamine Aminotransferase          87 non-null     float64
 6   Aspartate Aminotransferase        87 non-null     float64
 7   Total Protiens                    87 non-null     float64
 8   Albumin                           87 non-null     float64
 9   Ratio Albumin and Globulin Ratio  87 non-null     float64
dtypes: float64(10)
memory usage: 7.5 KB


# Over Sampling with SMOTE

In [13]:
sm = SMOTE(random_state=1)

X_train_smote, y_train_smote = sm.fit_sample(X_train, y_train)

# Training input data after SMOTE.
# Shape gives (row, columns).
print('The shape of train_X: {}'.format(X_train_smote.shape)) 
print('The shape of train_y: {} \n'.format(y_train_smote.shape)) 
  
# Training target data after SMOTE. 
# Note! The number of labels should be the same for both classes!
print("Number of liver patients '0': {}".format(sum(y_train_smote == 0))) 
print("Number of non-liver patients '1': {}".format(sum(y_train_smote == 1))) 

The shape of train_X: (570, 10)
The shape of train_y: (570,) 

Number of liver patients '0': 285
Number of non-liver patients '1': 285


In [14]:
X_train_smote.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  56.0     2.0              0.7               0.1                 145.0   
1  46.0     1.0              0.6               0.2                 115.0   
2  30.0     1.0              1.6               0.4                 332.0   
3  32.0     2.0              0.7               0.1                 240.0   
4  26.0     2.0              0.9               0.2                 154.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      26.0                        23.0             7.0   
1                      14.0                        11.0             6.9   
2                      84.0                       139.0             5.6   
3                      12.0                        15.0             7.0   
4                      16.0                        12.0             7.0   

   Albumin  Ratio Albumin and Globulin Ratio  
0      4.0                               1.3  
1      3.4                               0.9  
2      2.7                               0.9  
3      3.0                               0.7  
4      3.5                               1.0

In [15]:
y_train_smote.head()

0    1.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Liver Patient, dtype: float64

In [16]:
liverPatientDataSmote = X_train_smote

liverPatientDataSmote['Liver Patient'] = y_train_smote

liverPatientDataSmote.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  56.0     2.0              0.7               0.1                 145.0   
1  46.0     1.0              0.6               0.2                 115.0   
2  30.0     1.0              1.6               0.4                 332.0   
3  32.0     2.0              0.7               0.1                 240.0   
4  26.0     2.0              0.9               0.2                 154.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      26.0                        23.0             7.0   
1                      14.0                        11.0             6.9   
2                      84.0                       139.0             5.6   
3                      12.0                        15.0             7.0   
4                      16.0                        12.0             7.0   

   Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0      4.0                               1.3            1.0  
1      3.4                               0.9            0.0  
2      2.7                               0.9            0.0  
3      3.0                               0.7            0.0  
4      3.5                               1.0            0.0

In [17]:
liverPatientDataSmote.tail()

Age    Gender  Total Bilirubin  Direct Bilirubin  \
565  23.858531  1.000000         0.723821          0.200000   
566  25.755105  1.000000         0.604442          0.104442   
567  17.843651  1.718783         0.812487          0.284365   
568  26.730179  1.000000         1.119054          0.359527   
569  33.316220  1.073123         0.721937          0.207312   

     Alkaline Phosphotase  Alamine Aminotransferase  \
565            135.000000                 27.714633   
566            183.621851                 89.800716   
567            140.219313                 18.562434   
568            179.539641                 66.777490   
569            189.146247                 22.000000   

     Aspartate Aminotransferase  Total Protiens   Albumin  \
565                   26.714633        6.590569  3.562032   
566                   52.111641        5.513325  2.304442   
567                   34.312699        6.271984  3.337566   
568                   40.888745        6.019054  3.165090   
569                   42.122520        7.429249  3.165811   

     Ratio Albumin and Globulin Ratio  Liver Patient  
565                          1.119105            1.0  
566                          0.700000            1.0  
567                          1.115320            1.0  
568                          1.046036            1.0  
569                          0.721937            1.0

## Note! The SMOTE training indices are from 1 to 570 in MATLAB.

In [18]:
liverPatientDataSmote.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age                               570 non-null    float64
 1   Gender                            570 non-null    float64
 2   Total Bilirubin                   570 non-null    float64
 3   Direct Bilirubin                  570 non-null    float64
 4   Alkaline Phosphotase              570 non-null    float64
 5   Alamine Aminotransferase          570 non-null    float64
 6   Aspartate Aminotransferase        570 non-null    float64
 7   Total Protiens                    570 non-null    float64
 8   Albumin                           570 non-null    float64
 9   Ratio Albumin and Globulin Ratio  570 non-null    float64
 10  Liver Patient                     570 non-null    float64
dtypes: float64(11)
memory usage: 49.1 KB


In [19]:
val_sorted = X_val
val_sorted['Liver Patient'] = y_val

val_sorted.reset_index(drop=True, inplace=True)

/opt/anaconda3/envs/myenv_conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [20]:
val_sorted.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  37.0     1.0              1.8               0.8                 145.0   
1  58.0     1.0              1.0               0.5                 158.0   
2  42.0     2.0              0.7               0.2                 152.0   
3  65.0     2.0              0.7               0.1                 187.0   
4  53.0     2.0              0.7               0.1                 182.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      62.0                        58.0             5.7   
1                      37.0                        43.0             7.2   
2                      35.0                        81.0             6.2   
3                      16.0                        18.0             6.8   
4                      20.0                        33.0             4.8   

   Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0      2.9                              1.00            0.0  
1      3.6                              1.00            0.0  
2      3.2                              1.06            0.0  
3      3.3                              0.90            0.0  
4      1.9                              0.60            0.0

In [21]:
val_sorted.tail()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
82  54.0     2.0             23.2              12.6                 574.0   
83  52.0     2.0              0.6               0.1                 194.0   
84  48.0     1.0              3.2               1.6                 257.0   
85  42.0     1.0              6.8               3.2                 630.0   
86  62.0     1.0              0.6               0.1                 160.0   

    Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
82                      43.0                        47.0             7.2   
83                      10.0                        12.0             6.9   
84                      33.0                       116.0             5.7   
85                      25.0                        47.0             6.1   
86                      42.0                       110.0             4.9   

    Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
82      3.5                              0.90            0.0  
83      3.3                              0.90            1.0  
84      2.2                              0.62            0.0  
85      2.3                              0.60            1.0  
86      2.6                              1.10            1.0

In [22]:
liverPatientDataSmote = liverPatientDataSmote.append(val_sorted, ignore_index=True)

## Checking to see if the validation set append correctly to the end of the liverPatientDataSmote dataframe.

In [23]:
liverPatientDataSmote.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  56.0     2.0              0.7               0.1                 145.0   
1  46.0     1.0              0.6               0.2                 115.0   
2  30.0     1.0              1.6               0.4                 332.0   
3  32.0     2.0              0.7               0.1                 240.0   
4  26.0     2.0              0.9               0.2                 154.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      26.0                        23.0             7.0   
1                      14.0                        11.0             6.9   
2                      84.0                       139.0             5.6   
3                      12.0                        15.0             7.0   
4                      16.0                        12.0             7.0   

   Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0      4.0                               1.3            1.0  
1      3.4                               0.9            0.0  
2      2.7                               0.9            0.0  
3      3.0                               0.7            0.0  
4      3.5                               1.0            0.0

In [24]:
liverPatientDataSmote.tail()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
652  54.0     2.0             23.2              12.6                 574.0   
653  52.0     2.0              0.6               0.1                 194.0   
654  48.0     1.0              3.2               1.6                 257.0   
655  42.0     1.0              6.8               3.2                 630.0   
656  62.0     1.0              0.6               0.1                 160.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
652                      43.0                        47.0             7.2   
653                      10.0                        12.0             6.9   
654                      33.0                       116.0             5.7   
655                      25.0                        47.0             6.1   
656                      42.0                       110.0             4.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
652      3.5                              0.90            0.0  
653      3.3                              0.90            1.0  
654      2.2                              0.62            0.0  
655      2.3                              0.60            1.0  
656      2.6                              1.10            1.0

In [25]:
# Row 570 should have an Age of 37.0 and row 656 should be 62.0.

liverPatientDataSmote.iloc[570:]

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
570  37.0     1.0              1.8               0.8                 145.0   
571  58.0     1.0              1.0               0.5                 158.0   
572  42.0     2.0              0.7               0.2                 152.0   
573  65.0     2.0              0.7               0.1                 187.0   
574  53.0     2.0              0.7               0.1                 182.0   
..    ...     ...              ...               ...                   ...   
652  54.0     2.0             23.2              12.6                 574.0   
653  52.0     2.0              0.6               0.1                 194.0   
654  48.0     1.0              3.2               1.6                 257.0   
655  42.0     1.0              6.8               3.2                 630.0   
656  62.0     1.0              0.6               0.1                 160.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
570                      62.0                        58.0             5.7   
571                      37.0                        43.0             7.2   
572                      35.0                        81.0             6.2   
573                      16.0                        18.0             6.8   
574                      20.0                        33.0             4.8   
..                        ...                         ...             ...   
652                      43.0                        47.0             7.2   
653                      10.0                        12.0             6.9   
654                      33.0                       116.0             5.7   
655                      25.0                        47.0             6.1   
656                      42.0                       110.0             4.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
570      2.9                              1.00            0.0  
571      3.6                              1.00            0.0  
572      3.2                              1.06            0.0  
573      3.3                              0.90            0.0  
574      1.9                              0.60            0.0  
..       ...                               ...            ...  
652      3.5                              0.90            0.0  
653      3.3                              0.90            1.0  
654      2.2                              0.62            0.0  
655      2.3                              0.60            1.0  
656      2.6                              1.10            1.0  

[87 rows x 11 columns]

## Note! The validation indices of our SMOTE dataset are from 571 to 657 in MATLAB.

In [26]:
test_sorted = X_test
test_sorted['Liver Patient'] = y_test

test_sorted.reset_index(drop=True, inplace=True)

/opt/anaconda3/envs/myenv_conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [27]:
test_sorted.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  60.0     1.0              0.5               0.1                 500.0   
1  68.0     1.0              0.7               0.1                 145.0   
2  46.0     1.0             18.4               8.5                 450.0   
3  55.0     1.0              0.7               0.2                 290.0   
4  31.0     1.0              0.6               0.1                 175.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      20.0                        34.0             5.9   
1                      20.0                        22.0             5.8   
2                     119.0                       230.0             7.5   
3                      53.0                        58.0             6.8   
4                      48.0                        34.0             6.0   

   Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0      1.6                              0.37            1.0  
1      2.9                              1.00            0.0  
2      3.3                              0.70            0.0  
3      3.4                              1.00            0.0  
4      3.7                              1.60            0.0

In [28]:
test_sorted.tail()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
82  60.0     1.0              2.3               0.6                 272.0   
83  35.0     1.0              0.9               0.2                 190.0   
84  53.0     2.0              0.9               0.2                 210.0   
85  60.0     1.0             19.6               9.5                 466.0   
86  38.0     2.0              0.6               0.1                 165.0   

    Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
82                      79.0                        51.0             6.6   
83                      25.0                        20.0             6.4   
84                      35.0                        32.0             8.0   
85                      46.0                        52.0             6.1   
86                      22.0                        34.0             5.9   

    Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
82      3.5                               1.1            0.0  
83      3.6                               1.2            1.0  
84      3.9                               0.9            1.0  
85      2.0                               0.4            0.0  
86      2.9                               0.9            1.0

In [29]:
liverPatientDataSmote = liverPatientDataSmote.append(test_sorted, ignore_index=True)

## Checking to see if the test set append correctly to the end of the liverPatientDataSmote dataframe.

In [30]:
liverPatientDataSmote.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  56.0     2.0              0.7               0.1                 145.0   
1  46.0     1.0              0.6               0.2                 115.0   
2  30.0     1.0              1.6               0.4                 332.0   
3  32.0     2.0              0.7               0.1                 240.0   
4  26.0     2.0              0.9               0.2                 154.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      26.0                        23.0             7.0   
1                      14.0                        11.0             6.9   
2                      84.0                       139.0             5.6   
3                      12.0                        15.0             7.0   
4                      16.0                        12.0             7.0   

   Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0      4.0                               1.3            1.0  
1      3.4                               0.9            0.0  
2      2.7                               0.9            0.0  
3      3.0                               0.7            0.0  
4      3.5                               1.0            0.0

In [31]:
liverPatientDataSmote.tail()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
739  60.0     1.0              2.3               0.6                 272.0   
740  35.0     1.0              0.9               0.2                 190.0   
741  53.0     2.0              0.9               0.2                 210.0   
742  60.0     1.0             19.6               9.5                 466.0   
743  38.0     2.0              0.6               0.1                 165.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
739                      79.0                        51.0             6.6   
740                      25.0                        20.0             6.4   
741                      35.0                        32.0             8.0   
742                      46.0                        52.0             6.1   
743                      22.0                        34.0             5.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
739      3.5                               1.1            0.0  
740      3.6                               1.2            1.0  
741      3.9                               0.9            1.0  
742      2.0                               0.4            0.0  
743      2.9                               0.9            1.0

In [33]:
# Row 657 should have an Age of 60.0 and row 743 should be 38.0.

liverPatientDataSmote.iloc[657:]

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
657  60.0     1.0              0.5               0.1                 500.0   
658  68.0     1.0              0.7               0.1                 145.0   
659  46.0     1.0             18.4               8.5                 450.0   
660  55.0     1.0              0.7               0.2                 290.0   
661  31.0     1.0              0.6               0.1                 175.0   
..    ...     ...              ...               ...                   ...   
739  60.0     1.0              2.3               0.6                 272.0   
740  35.0     1.0              0.9               0.2                 190.0   
741  53.0     2.0              0.9               0.2                 210.0   
742  60.0     1.0             19.6               9.5                 466.0   
743  38.0     2.0              0.6               0.1                 165.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
657                      20.0                        34.0             5.9   
658                      20.0                        22.0             5.8   
659                     119.0                       230.0             7.5   
660                      53.0                        58.0             6.8   
661                      48.0                        34.0             6.0   
..                        ...                         ...             ...   
739                      79.0                        51.0             6.6   
740                      25.0                        20.0             6.4   
741                      35.0                        32.0             8.0   
742                      46.0                        52.0             6.1   
743                      22.0                        34.0             5.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
657      1.6                              0.37            1.0  
658      2.9                              1.00            0.0  
659      3.3                              0.70            0.0  
660      3.4                              1.00            0.0  
661      3.7                              1.60            0.0  
..       ...                               ...            ...  
739      3.5                              1.10            0.0  
740      3.6                              1.20            1.0  
741      3.9                              0.90            1.0  
742      2.0                              0.40            0.0  
743      2.9                              0.90            1.0  

[87 rows x 11 columns]

## Note! The test indices of our SMOTE dataset are from 658 to 744 in MATLAB.

## Exporting SMOTE dataset as csv!

Some of the values have not rounded properly so we will do that manually now.

In [55]:
liverPatientDataSmote.Age = liverPatientDataSmote.Age.round()
liverPatientDataSmote.Gender = liverPatientDataSmote.Gender.round()
liverPatientDataSmote['Total Bilirubin'] = liverPatientDataSmote['Total Bilirubin'].round(1)
liverPatientDataSmote['Direct Bilirubin'] = liverPatientDataSmote['Direct Bilirubin'].round(1)
liverPatientDataSmote['Alkaline Phosphotase'] = liverPatientDataSmote['Alkaline Phosphotase'].round()
liverPatientDataSmote['Alamine Aminotransferase'] = liverPatientDataSmote['Alamine Aminotransferase'].round()
liverPatientDataSmote['Aspartate Aminotransferase'] = liverPatientDataSmote['Aspartate Aminotransferase'].round()
liverPatientDataSmote['Total Protiens'] = liverPatientDataSmote['Total Protiens'].round(1)
liverPatientDataSmote['Albumin'] = liverPatientDataSmote['Albumin'].round(1)
liverPatientDataSmote['Ratio Albumin and Globulin Ratio'] = liverPatientDataSmote['Ratio Albumin and Globulin Ratio'].round(1)

In [56]:
liverPatientDataSmote.to_csv('../MATLAB Code/liverPatientDataSmote.csv', index=False)

# Under Sampling with NearMiss

In [58]:
nm = NearMiss() 
X_train_miss, y_train_miss = nm.fit_sample(X_train, y_train.ravel()) 

# Training input data after NearMiss.
# Shape gives (row, columns).
print('The shape of train_X: {}'.format(X_train_miss.shape)) 
print('The shape of train_y: {} \n'.format(y_train_miss.shape)) 
  
# Training target data after NearMiss. 
# Note! The number of labels should be the same for both classes!
print("Number of liver patients '0': {}".format(sum(y_train_miss == 0))) 
print("Number of non-liver patients '1': {}".format(sum(y_train_miss == 1))) 

The shape of train_X: (240, 10)
The shape of train_y: (240,) 

Number of liver patients '0': 120
Number of non-liver patients '1': 120


In [59]:
liverPatientDataMiss = X_train_miss

liverPatientDataMiss['Liver Patient'] = y_train_miss

liverPatientDataMiss.head()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
0  64.0     1.0              1.1               0.4                 201.0   
1  55.0     1.0              0.9               0.2                 190.0   
2  51.0     1.0              0.7               0.1                 180.0   
3  45.0     1.0              2.5               1.2                 163.0   
4  33.0     1.0              1.8               0.8                 196.0   

   Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
0                      18.0                        19.0             6.9   
1                      25.0                        28.0             5.9   
2                      25.0                        27.0             6.1   
3                      28.0                        22.0             7.6   
4                      25.0                        22.0             8.0   

   Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
0      4.1                               1.4            0.0  
1      2.7                               0.8            0.0  
2      3.1                               1.0            0.0  
3      4.0                               1.1            0.0  
4      4.0                               1.0            0.0

In [61]:
liverPatientDataMiss.tail()

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
235  50.0     2.0              1.0               0.3                 191.0   
236  42.0     1.0              0.7               0.2                 197.0   
237  65.0     2.0              0.7               0.2                 182.0   
238  56.0     1.0              1.0               0.3                 195.0   
239  22.0     1.0              0.8               0.2                 300.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
235                      22.0                        31.0             7.8   
236                      64.0                        33.0             5.8   
237                      23.0                        28.0             6.8   
238                      22.0                        28.0             5.8   
239                      57.0                        40.0             7.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
235      4.0                               1.0            1.0  
236      2.4                               0.7            1.0  
237      2.9                               0.7            1.0  
238      2.6                               0.8            1.0  
239      3.8                               0.9            1.0

In [60]:
liverPatientDataMiss.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age                               240 non-null    float64
 1   Gender                            240 non-null    float64
 2   Total Bilirubin                   240 non-null    float64
 3   Direct Bilirubin                  240 non-null    float64
 4   Alkaline Phosphotase              240 non-null    float64
 5   Alamine Aminotransferase          240 non-null    float64
 6   Aspartate Aminotransferase        240 non-null    float64
 7   Total Protiens                    240 non-null    float64
 8   Albumin                           240 non-null    float64
 9   Ratio Albumin and Globulin Ratio  240 non-null    float64
 10  Liver Patient                     240 non-null    float64
dtypes: float64(11)
memory usage: 20.8 KB


## Note! The NearMiss training indices are from 1 to 240 in MATLAB.

In [62]:
liverPatientDataMiss = liverPatientDataMiss.append(val_sorted, ignore_index=True)

liverPatientDataMiss.iloc[240:]

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
240  37.0     1.0              1.8               0.8                 145.0   
241  58.0     1.0              1.0               0.5                 158.0   
242  42.0     2.0              0.7               0.2                 152.0   
243  65.0     2.0              0.7               0.1                 187.0   
244  53.0     2.0              0.7               0.1                 182.0   
..    ...     ...              ...               ...                   ...   
322  54.0     2.0             23.2              12.6                 574.0   
323  52.0     2.0              0.6               0.1                 194.0   
324  48.0     1.0              3.2               1.6                 257.0   
325  42.0     1.0              6.8               3.2                 630.0   
326  62.0     1.0              0.6               0.1                 160.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
240                      62.0                        58.0             5.7   
241                      37.0                        43.0             7.2   
242                      35.0                        81.0             6.2   
243                      16.0                        18.0             6.8   
244                      20.0                        33.0             4.8   
..                        ...                         ...             ...   
322                      43.0                        47.0             7.2   
323                      10.0                        12.0             6.9   
324                      33.0                       116.0             5.7   
325                      25.0                        47.0             6.1   
326                      42.0                       110.0             4.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
240      2.9                              1.00            0.0  
241      3.6                              1.00            0.0  
242      3.2                              1.06            0.0  
243      3.3                              0.90            0.0  
244      1.9                              0.60            0.0  
..       ...                               ...            ...  
322      3.5                              0.90            0.0  
323      3.3                              0.90            1.0  
324      2.2                              0.62            0.0  
325      2.3                              0.60            1.0  
326      2.6                              1.10            1.0  

[87 rows x 11 columns]

## Note! The NearMiss validation indices are from 241 to 327 in MATLAB.

In [64]:
liverPatientDataMiss = liverPatientDataMiss.append(test_sorted, ignore_index=True)

liverPatientDataMiss.iloc[327:]

Age  Gender  Total Bilirubin  Direct Bilirubin  Alkaline Phosphotase  \
327  60.0     1.0              0.5               0.1                 500.0   
328  68.0     1.0              0.7               0.1                 145.0   
329  46.0     1.0             18.4               8.5                 450.0   
330  55.0     1.0              0.7               0.2                 290.0   
331  31.0     1.0              0.6               0.1                 175.0   
..    ...     ...              ...               ...                   ...   
409  60.0     1.0              2.3               0.6                 272.0   
410  35.0     1.0              0.9               0.2                 190.0   
411  53.0     2.0              0.9               0.2                 210.0   
412  60.0     1.0             19.6               9.5                 466.0   
413  38.0     2.0              0.6               0.1                 165.0   

     Alamine Aminotransferase  Aspartate Aminotransferase  Total Protiens  \
327                      20.0                        34.0             5.9   
328                      20.0                        22.0             5.8   
329                     119.0                       230.0             7.5   
330                      53.0                        58.0             6.8   
331                      48.0                        34.0             6.0   
..                        ...                         ...             ...   
409                      79.0                        51.0             6.6   
410                      25.0                        20.0             6.4   
411                      35.0                        32.0             8.0   
412                      46.0                        52.0             6.1   
413                      22.0                        34.0             5.9   

     Albumin  Ratio Albumin and Globulin Ratio  Liver Patient  
327      1.6                              0.37            1.0  
328      2.9                              1.00            0.0  
329      3.3                              0.70            0.0  
330      3.4                              1.00            0.0  
331      3.7                              1.60            0.0  
..       ...                               ...            ...  
409      3.5                              1.10            0.0  
410      3.6                              1.20            1.0  
411      3.9                              0.90            1.0  
412      2.0                              0.40            0.0  
413      2.9                              0.90            1.0  

[87 rows x 11 columns]

## Note! The NearMiss testing indices are from 328 to 414 in MATLAB.

In [66]:
liverPatientDataMiss.to_csv('../MATLAB Code/liverPatientDataMiss.csv', index=False)